### ISSUE TITLE
run Gratipay 154

### ISSUE THREAD

**Issue by whit537** [gratipay/gratipay.com#3415]
[153](https://github.com/gratipay/gratipay.com/issues/3385) was the [Gratipocalypse](https://medium.com/gratipay-blog/gratipocalypse-42fd0ec0d9e8). See #3399 for our efforts to get back on track for 154.

**Comment by whit537** [comment]
I believe the payday rewrite is far enough along for this week (#3414).

**Comment by whit537** [comment]
I'm going to migrate `tips` to `subscriptions` on #3399, and then switch over to here.

**Comment by whit537** [comment]
```
=> select count(*) from subscriptions;
┌───────┐
│ count │
├───────┤
│   984 │
└───────┘
(1 row)
```

**Comment by whit537** [comment]
Who's ready?

**Comment by whit537** [comment]
Account review performed manually for beta users of the new system (see #3399).

**Comment by whit537** [comment]
Droplet spun up, backup started ...

**Comment by rohitpaulk** [comment]
:clap: :clap:

**Comment by whit537** [comment]
Droplet updated.

**Comment by whit537** [comment]
Backup finished and (manually) verified.

**Comment by whit537** [comment]
Script started.

**Comment by whit537** [comment]
Seems to be working. :four_leaf_clover:

**Comment by tshepang** [comment]
yeah, just got notification sms

**Comment by whit537** [comment]
`Captured 110 card holds.`

**Comment by whit537** [comment]
@tshepang Nice!

The script seems to have paused ...

**Comment by whit537** [comment]
Could be because we're canceling so many old card holds?

**Comment by whit537** [comment]
I still see the process:

```
$ ps auxww|grep python
whit537   1504  0.0  2.1  45592 10928 ?        S    21:34   0:00 /home/whit537/gratipay.com/env/bin/python2.7 ./env/bin/honcho run -e /dev/stdin ./env/bin/payday
whit537   1524  1.1  7.5 743124 37988 ?        Sl   21:34   0:09 /home/whit537/gratipay.com/env/bin/python2.7 ./env/bin/payday
```

**Comment by whit537** [comment]
Must be stuck in card holds, because the next log line is `Canceled %i card holds.`, and I haven't seen that yet.

**Comment by whit537** [comment]
I guess I'll give it a couple more minutes? What's our option, comment that out and rerun the script?

**Comment by whit537** [comment]
There's a step at the beginning of payday that takes quite a bit of time, iirc it's when we fetch card holds to begin with (it comes after the db is prepared but before we start placing new holds).

**Comment by rohitpaulk** [comment]
> What's our option, comment that out and rerun the script?

We could add a log message to `cancel_card_hold` just like `create_card_hold` so that we're sure that's where it's stuck.

**Comment by whit537** [comment]
Killed it.

**Comment by whit537** [comment]
Put an early `return` in `settle_card_holds` and restarted ...

**Comment by whit537** [comment]
Okay, commented out the parts we already ran and re-ran. We are making many more payouts than I intended. Oh well.

**Comment by whit537** [comment]
Did payout for 354 participants.
Script ran for 35 seconds (0:00:35.063665).

**Comment by whit537** [comment]
Okay!

**Comment by rohitpaulk** [comment]
> Could be because we're canceling so many old card holds?

Canceling old card holds would happen early on in `fetch_card_holds`, so can't be that...

**Comment by whit537** [comment]
Log downloaded and droplet destroyed. I will pick up with this again later, I have to go for now ...

You all are wonderful. :-)

**Comment by rohitpaulk** [comment]
The `payments` table is empty, reticketing...

**Comment by rohitpaulk** [comment]
This doesn't look good. 

I only see 8 exchanges (amount < 0), and none of those are owners of the teams that were supposed to take part in payday today.

**Comment by rohitpaulk** [comment]
> Put an early return in settle_card_holds and restarted

@whit537 - I know this is highly unlikely, but did you happen to put the early return in between `self.settle_card_holds(cursor, holds)` and `self.update_balances(cursor)` and not within `settle_card_holds`?

From a quick look, that's the only possibility I see.

**Comment by rohitpaulk** [comment]
It's 4 AM here, I'm off to sleep. Will dig further as soon as I can.

**Comment by rohitpaulk** [comment]
A note for anyone looking at this ticket and concerned : 

We've pulled in money from CCs and stored it as balance in user's accounts. From the looks of it, all that we missed is the the internal shuffling of balances(from participant to teams).

**Comment by rohitpaulk** [comment]
@whit537 - Here's what I'm thinking. We edit payday 154 manually, mark it as not completed by modifying `ts_end`. And then, we make the following changes to payday.py and run it - 

- run
  - payin
      - prepare
      - create_card_holds (**Short circuit this, return an empty list**)
      - process_subscriptions
      - transfer_takes
      - process_draws
      - settle_card_holds (**Since holds is empty, this won't do anything.**)
      - update_balances
      - take_over_balances
  - payout (**Constrain this to the new teams only**)
  - update_stats
  - update_cached_amounts
  - end
```

**Comment by rohitpaulk** [comment]
And once we're done with that, or even before - we cancel all the pending holds we've created.

**Comment by rohitpaulk** [comment]
Blocked on #3430. It'd be good to merge #3431 too before rerunning this, but it's not necessary.

**Comment by whit537** [comment]
Okay, I've made it through my GitHub notifications and I'm topping up support. I'll be back here in a few minutes ...

**Comment by rohitpaulk** [comment]
> I know this is highly unlikely, but did you happen to put the early return in between self.settle_card_holds(cursor, holds) and self.update_balances(cursor) and not within settle_card_holds?

I had a look at the logs, and it's definitely not that. 

@whit537 - What do you think happened in https://github.com/gratipay/logs/blob/master/payday/gratipay-154.log#L354-L393?

**Comment by whit537** [comment]
>What do you think happened in https://github.com/gratipay/logs/blob/master/payday/gratipay-154.log#L354-L393?

I almost saved a diff of my hotfix to payday, but I didn't. Sorry. :-(

In fact, I ran it with several hotfixes. Here's what I ended up doing, as best I remember:

 1. I added a `return` just before [canceling card holds](https://github.com/gratipay/gratipay.com/blob/1839/gratipay/billing/payday.py#L332).
 1. I reran payday. It was spinning again. As I recall, I had the impression that it was spinning in `fetch_card_holds`, which is normal behavior as mentioned above at https://github.com/gratipay/gratipay.com/issues/3415#issuecomment-102180134.
 1. I killed it.
 1. In `payin`, I commented out [lines 161-165](https://github.com/gratipay/gratipay.com/blob/1839/gratipay/billing/payday.py#L161-L165) as well as line 170. I really don't think I commented out line 171, the called to `update_balances`, but without the diff I suppose that's a mistake I may have made.
 1. I reran payday. It failed because I had commented out `prepare`.
 1. I uncommented out `prepare` and reran. It completed.

**Comment by whit537** [comment]
I see 8 payouts and 110 charges in the Balanced dashboard, which lines up with `exchanges`. The log reported 110 card captures, so that's fine. Why does the log say "Did payout for 354 participants." when we only did 8?

**Comment by rohitpaulk** [comment]
Because only 8 had balances more than 10, is what I think.

**Comment by whit537** [comment]
Yeah, we tighten up [that log line](https://github.com/gratipay/gratipay.com/blob/1839/gratipay/billing/payday.py#L433).

**Comment by rohitpaulk** [comment]
> Minimum payout is $10.00. XXX is only due $6.00.

**Comment by whit537** [comment]
> Yeah, we should tighten up [that log line](https://github.com/gratipay/gratipay.com/blob/1839/gratipay/billing/payday.py#L433).

Reticketed as #3435.

**Comment by rohitpaulk** [comment]
> In payin, I commented out lines 161-165 as well as line 170

Looks like that's where the mistake is. :)

> I really don't think I commented out line 171, the called to update_balances

The call to update_balances worked, just that we didn't update any balances, coz `payday_participants` was empty. 

> https://github.com/gratipay/logs/blob/master/payday/gratipay-154.log#L392
> Updated the balances of 0 participants.

Hmm, starting to make sense now...

**Comment by rohitpaulk** [comment]
I think that https://github.com/gratipay/gratipay.com/issues/3415#issuecomment-102216057 should work... 

What say, @whit537?

It's basically doing an internal shuffle, without touching CCs. And we should comment out `notify_participants` too, we don't want to send duplicate emails (again :P).

**Comment by whit537** [comment]
> ... coz `payday_participants` was empty.

And `payday_participants` was empty, because ... I had commented out `prepare`? But I *un*commented `prepare` in step (6).

**Comment by rohitpaulk** [comment]
Sorry, my mistake. 

`payday_participants` was populated but it wasn't modified (in `process_draws` and `transfer_takes`). (new_balance and old_balance were the same).

**Comment by whit537** [comment]
Gotcha. `transfer_takes` is a [noop](https://github.com/gratipay/gratipay.com/blob/1839/gratipay/billing/payday.py#L286) this week ([ref](https://github.com/gratipay/gratipay.com/pull/3414#issuecomment-102130834)), but we needed to run `process_draws`, yes. We would've needed to run `process_subscriptions`, too, for that matter, because that's where we populate the team balances that we than redistribute back to participants in `_draws`.

**Comment by whit537** [comment]
So the reason there are no `payments` is because there were no payments. :-)

**Comment by rohitpaulk** [comment]
I'm sorry for the rough explanation, I was on the phone :)

> So the reason there are no payments is because there were no payments. :-)

Yep.

**Comment by whit537** [comment]
> What say, @whit537?

Yeah, something like https://github.com/gratipay/gratipay.com/issues/3415#issuecomment-102216057 sounds about right. I'm spinning up another droplet, I'll paste a diff here before running. We should probably handle the payout constraint with a PR prior to rerunning, since that's a permanent change we want to make.

**Comment by whit537** [comment]
> handle the payout constraint

How about we constrain to participants that are in `teams.owner` or `payroll.member`?

**Comment by rohitpaulk** [comment]
> How about we constrain to participants that are in teams.owner or payroll.member?

Yep, sounds right.

**Comment by rohitpaulk** [comment]
I'll handle the constraint, you can go ahead with the patching.

**Comment by whit537** [comment]
k

**Comment by whit537** [comment]
Droplet spun up logfile put back in place.

**Comment by whit537** [comment]
Droplet updated and built.

**Comment by whit537** [comment]
Why did we payout the 8 people we *did* payout to?

**Comment by rohitpaulk** [comment]
Because they had balances greater than the minimum credit amount **and** they had a working payout route.

**Comment by whit537** [comment]
> We edit payday 154 manually, mark it as not completed by modifying `ts_end`.

Done:

> => update paydays set ts_end='1970-01-01T00:00:00+00'::timestamptz where id=199;

**Comment by rohitpaulk** [comment]
I'm working on #3438, almost done.

**Comment by rohitpaulk** [comment]
@whit537 - Need your help finishing up #3438, look at https://github.com/gratipay/gratipay.com/commit/1e069e88def0a38dc0d2322bff05a9a9bf2e4ad0

**Comment by whit537** [comment]
k

**Comment by whit537** [comment]
How's this look, @rohitpaulk?

```diff
diff --git a/gratipay/billing/payday.py b/gratipay/billing/payday.py
index c8513fc..ac9b75b 100644
--- a/gratipay/billing/payday.py
+++ b/gratipay/billing/payday.py
@@ -145,7 +145,7 @@ class Payday(object):
             self.mark_stage_done()

         self.end()
-        self.notify_participants()
+        #self.notify_participants()

         _end = aspen.utils.utcnow()
         _delta = _end - _start
@@ -159,7 +159,7 @@ class Payday(object):
         """
         with self.db.get_cursor() as cursor:
             self.prepare(cursor, self.ts_start)
-            holds = self.create_card_holds(cursor)
+            #holds = self.create_card_holds(cursor)
             self.process_subscriptions(cursor)
             self.transfer_takes(cursor, self.ts_start)
             self.process_draws(cursor)
@@ -167,7 +167,7 @@ class Payday(object):
                 SELECT * FROM payments WHERE "timestamp" > %s
             """, (self.ts_start,))
             try:
-                self.settle_card_holds(cursor, holds)
+                self.settle_card_holds(cursor, {})
                 self.update_balances(cursor)
                 check_db(cursor)
             except:
```

**Comment by rohitpaulk** [comment]
Looks good.

**Comment by rohitpaulk** [comment]
Updated the droplet with the latest changes?

**Comment by whit537** [comment]
Yup, getting ready ...

**Comment by whit537** [comment]
Um ...

```
Fri May 15 18:39:07 UTC 2015
pid-2918 thread-140194056021760 (MainThread) Picking up with an existing payday.
pid-2918 thread-140194056021760 (MainThread) Payday started at 2015-05-14 21:34:37.995735+00:00.
pid-2918 thread-140194056021760 (MainThread) Greetings, program! It's PAYDAY!!!!
pid-2918 thread-140194056021760 (MainThread) Script ran for just a moment (0:00:00.029568).
```

**Comment by rohitpaulk** [comment]
ah, we didn't modify the `stage`

**Comment by whit537** [comment]
Good call.

**Comment by rohitpaulk** [comment]
Change it to.... 0, i guess? Yh, 0.

**Comment by whit537** [comment]
```
=> update paydays set stage=0 where id=199;
```

**Comment by whit537** [comment]
Erm, we also needed to reset `ts_end` again. I think I caught it in time:

```
Fri May 15 18:41:48 UTC 2015
pid-3033 thread-139956672141056 (MainThread) Starting a new payday.
pid-3033 thread-139956672141056 (MainThread) Payday started at 2015-05-15 18:42:11.415296+00:00.
pid-3033 thread-139956672141056 (MainThread) Greetings, program! It's PAYDAY!!!!
pid-3033 thread-139956672141056 (MainThread) Prepared the DB.
Terminated
```

**Comment by whit537** [comment]
`payments` is still empty, and `exchanges` doesn't have any new records.

**Comment by rohitpaulk** [comment]
> I think I caught it in time

:relieved:

**Comment by whit537** [comment]
I think we're good. So:

 - [ ] delete the new payday I accidentally started
 - [ ] reset `ts_end` of payday `199`
 - [ ] stage is already at `0`

**Comment by whit537** [comment]
Right?

**Comment by rohitpaulk** [comment]
Yep...

**Comment by whit537** [comment]
```
Fri May 15 18:49:04 UTC 2015
pid-3176 thread-139946617497344 (MainThread) Picking up with an existing payday.
pid-3176 thread-139946617497344 (MainThread) Payday started at 2015-05-14 21:34:37.995735+00:00.
pid-3176 thread-139946617497344 (MainThread) Greetings, program! It's PAYDAY!!!!
pid-3176 thread-139946617497344 (MainThread) Prepared the DB.
```

**Comment by rohitpaulk** [comment]
We haven't updated `update_stats`, guess that can wait... 

I wonder whether fake payday is what it's holding up on.

**Comment by rohitpaulk** [comment]
> I wonder whether fake payday is what it's holding up on. 

Nah, don't think so. It's independent of payday, and we haven't made any changes to it.

**Comment by whit537** [comment]
Yeah, I expect the stats will be off. I figured that was lower priority.

**Comment by rohitpaulk** [comment]
True.

**Comment by whit537** [comment]
Process is still blocking ...

**Comment by rohitpaulk** [comment]
What's the last log line?

**Comment by rohitpaulk** [comment]
Don't terminate it just yet... Let's see if we can figure out what's causing the block this time.

**Comment by whit537** [comment]
> What's the last log line?

`Prepared the DB.`

> Don't terminate it just yet... Let's see if we can figure out what's causing the block this time.

It's not pegging CPU/RAM, which indicates against an infinite loop.

**Comment by whit537** [comment]
Does Balanced give us an API call log? That could give us some useful info ...

**Comment by rohitpaulk** [comment]
We aren't making any API calls in this one..

**Comment by whit537** [comment]
Well, not until payouts. Balanced does have a log, btw, but yeah there's relevant there.

**Comment by rohitpaulk** [comment]
@whit537 - Looks like we haven't done any damage (yet). Let's add a few more log statements? To `process_subscriptions`, `process_draws` etc.

**Comment by rohitpaulk** [comment]
Hmm, we should be able to inspect the tables too.

**Comment by rohitpaulk** [comment]
Or not?

**Comment by rohitpaulk** [comment]
Only visible within the transaction. :(

**Comment by whit537** [comment]
Right. We could take out "ON COMMIT DELETE" and maybe they'd stick around? We could delete them explicitly after payday is done.

**Comment by rohitpaulk** [comment]
> A temporary table is only available and accessible by the client who creates the table.

We'll have to take out the `TEMPORARY TABLE` too.

**Comment by rohitpaulk** [comment]
Let's do that, **and** add simple log statements to `process_subscriptions` and `process_draws`. Good enough for a next try?

**Comment by rohitpaulk** [comment]
We should do this on a backup :/

**Comment by rohitpaulk** [comment]
Once we're sure that we're reaching the payout stage, we could run on production.

**Comment by whit537** [comment]
Alright, we've got #3440 and #3441 on master. Stage is still `0` and `ts_end` is still `1970` for `199`. Time to update the droplet and rerun ...

**Comment by whit537** [comment]
To be clear, we're expecting it to block again, but this time we'll have more info for debugging.

**Comment by whit537** [comment]
I did reapply the diff: https://github.com/gratipay/gratipay.com/issues/3415#issuecomment-102486250.

**Comment by whit537** [comment]
Interesting:

```
Fri May 15 19:33:16 UTC 2015
pid-3572 thread-140704402380544 (MainThread) Picking up with an existing payday.
pid-3572 thread-140704402380544 (MainThread) Payday started at 2015-05-14 21:34:37.995735+00:00.
pid-3572 thread-140704402380544 (MainThread) Greetings, program! It's PAYDAY!!!!
pid-3572 thread-140704402380544 (MainThread) Prepared the DB.
pid-3572 thread-140704402380544 (MainThread) Processing subscriptions.
pid-3572 thread-140704402380544 (MainThread) Processing draws.
pid-3572 thread-140704402380544 (MainThread) Settling card holds.
```

**Comment by rohitpaulk** [comment]
Hmm.... :question:

**Comment by ehmatthes** [comment]
!m you two!

**Comment by rohitpaulk** [comment]
@ehmatthes - :)

**Comment by rohitpaulk** [comment]
Something to do with threaded_map, @whit537 ?

**Comment by whit537** [comment]
```
Fri May 15 19:50:26 UTC 2015
pid-3964 thread-140369408612096 (MainThread) Picking up with an existing payday.
pid-3964 thread-140369408612096 (MainThread) Payday started at 2015-05-14 21:34:37.995735+00:00.
pid-3964 thread-140369408612096 (MainThread) Greetings, program! It's PAYDAY!!!!
pid-3964 thread-140369408612096 (MainThread) Prepared the DB.
pid-3964 thread-140369408612096 (MainThread) Processing subscriptions.
pid-3964 thread-140369408612096 (MainThread) Processing draws.
pid-3964 thread-140369408612096 (MainThread) Settling card holds.
pid-3964 thread-140369408612096 (MainThread) Settling card holds for 0 participants.
pid-3964 thread-140369408612096 (MainThread) Capturing card holds.
```

**Comment by whit537** [comment]
@rohitpaulk Could be ...

**Comment by rohitpaulk** [comment]
I can't access the tables via psql. Are we running this on a backup?

**Comment by whit537** [comment]
No, not running on a backup. I'm rerunning with `threads=1` in the `threaded_map` signature. That's helped us in the past.

**Comment by whit537** [comment]
Still blocking at `Capturing card holds.`.

**Comment by rohitpaulk** [comment]
```
pid-21713 thread-140077650589440 (MainThread) Picking up with an existing payday.
pid-21713 thread-140077650589440 (MainThread) Payday started at 2015-05-14 21:34:37.995735+00:00.
pid-21713 thread-140077650589440 (MainThread) Greetings, program! It's PAYDAY!!!!
pid-21713 thread-140077650589440 (MainThread) Prepared the DB.
pid-21713 thread-140077650589440 (MainThread) Processing subscriptions.
pid-21713 thread-140077650589440 (MainThread) Processing draws.
pid-21713 thread-140077650589440 (MainThread) Settling card holds.
pid-21713 thread-140077650589440 (MainThread) Capturing card holds.
pid-21713 thread-140077650589440 (MainThread) Captured 0 card holds.
pid-21713 thread-140077650589440 (MainThread) Canceling card holds.
pid-21713 thread-140077650589440 (MainThread) Canceled 0 card holds.
pid-21713 thread-140077650589440 (MainThread) Updating balances.
pid-21713 thread-140077650589440 (MainThread) Updated the balances of 639 participants.
pid-21713 thread-140077650589440 (MainThread) Taking over balances.
```

Seems to work for me on a backup... Haven't done anything different, as far as I can tell.

**Comment by whit537** [comment]
@rohitpaulk That's using master with https://github.com/gratipay/gratipay.com/issues/3415#issuecomment-102486250?

**Comment by rohitpaulk** [comment]
Yep, master, and the difff is pretty much the same thing.

```diff
diff --git a/gratipay/billing/payday.py b/gratipay/billing/payday.py
index 672a92f..3fbf807 100644
--- a/gratipay/billing/payday.py
+++ b/gratipay/billing/payday.py
@@ -137,7 +137,8 @@ class Payday(object):
             self.payin()
             self.mark_stage_done()
         if self.stage < 2:
-            self.payout()
+            raise ValueError
+            #self.payout()
             self.mark_stage_done()
         if self.stage < 3:
             self.update_stats()
@@ -145,7 +146,7 @@ class Payday(object):
             self.mark_stage_done()
 
         self.end()
-        self.notify_participants()
+        #self.notify_participants()
 
         _end = aspen.utils.utcnow()
         _delta = _end - _start
@@ -159,7 +160,7 @@ class Payday(object):
         """
         with self.db.get_cursor() as cursor:
             self.prepare(cursor, self.ts_start)
-            holds = self.create_card_holds(cursor)
+            #holds = self.create_card_holds(cursor)
             self.process_subscriptions(cursor)
             self.transfer_takes(cursor, self.ts_start)
             self.process_draws(cursor)
@@ -167,7 +168,7 @@ class Payday(object):
                 SELECT * FROM payments WHERE "timestamp" > %s
             """, (self.ts_start,))
             try:
-                self.settle_card_holds(cursor, holds)
+                self.settle_card_holds(cursor, {})
                 self.update_balances(cursor)
                 check_db(cursor)
             except:
```

**Comment by whit537** [comment]
```
Fri May 15 20:12:35 UTC 2015
pid-2887 thread-140735252603648 (MainThread) Picking up with an existing payday.
pid-2887 thread-140735252603648 (MainThread) Payday started at 2015-05-14 21:34:37.995735+00:00.
pid-2887 thread-140735252603648 (MainThread) Greetings, program! It's PAYDAY!!!!
pid-2887 thread-140735252603648 (MainThread) Prepared the DB.
pid-2887 thread-140735252603648 (MainThread) Processing subscriptions.
pid-2887 thread-140735252603648 (MainThread) Processing draws.
pid-2887 thread-140735252603648 (MainThread) Settling card holds.
pid-2887 thread-140735252603648 (MainThread) Capturing card holds.
pid-2887 thread-140735252603648 (MainThread) Captured 0 card holds.
pid-2887 thread-140735252603648 (MainThread) Canceling card holds.
pid-2887 thread-140735252603648 (MainThread) Canceled 0 card holds.
pid-2887 thread-140735252603648 (MainThread) Updating balances.
pid-2887 thread-140735252603648 (MainThread) Updated the balances of 639 participants.
pid-2887 thread-140735252603648 (MainThread) Taking over balances.
pid-2887 thread-140735252603648 (MainThread) Starting payout loop.
[...]
pid-2887 thread-140735252603648 (MainThread) Did payout for 4 participants.
pid-2887 thread-140735252603648 (MainThread) Checked the DB.
pid-2887 thread-140735252603648 (MainThread) Updating stats.
pid-2887 thread-140735252603648 (MainThread) Updated payday stats.
pid-2887 thread-140735252603648 (MainThread) Updated receiving amounts.
pid-2887 thread-140735252603648 (MainThread) Script ran for 18 seconds (0:00:18.109225).
```

**Comment by whit537** [comment]
Phew! O.o

You gave me the idea of running it on my laptop instead of on a droplet. That worked. :)

**Comment by whit537** [comment]
Okay! Where are we?

**Comment by rohitpaulk** [comment]
Ha! :dancer: :dancer:

**Comment by whit537** [comment]
What did we break? :) :hammer:

**Comment by whit537** [comment]
```
=> select count(*) from payments;
┌───────┐
│ count │
├───────┤
│   661 │
└───────┘
(1 row)
```

Yesssssssss.

**Comment by rohitpaulk** [comment]
Is this one of those python version mismatch bugs like last time? Just to make sure, my `./env/bin/python` is `2.7.8`.

**Comment by whit537** [comment]
We made three new exchanges. We have six teams, two of which are on PayPal. The sixth had a balance less than $10. I've mentioned the payout threshold at https://github.com/gratipay/gratipay.com/issues/1383#issuecomment-102516158.

**Comment by whit537** [comment]
Gosh, maybe? Do you have the ticket handy for "last time"? I'm on 2.7.9 and the droplet is on 2.7.3.

**Comment by rohitpaulk** [comment]
Last time was Honcho... 

https://github.com/gratipay/gratipay.com/issues/3025
https://github.com/gratipay/gratipay.com/pull/3028

**Comment by rohitpaulk** [comment]
Finally.....

**Comment by whit537** [comment]
Not so fast. We still need to do MassPay and shuffle escrow. :package:

**Comment by whit537** [comment]
(I was delayed by a phone call re: #3366, sorry.)

**Comment by whit537** [comment]
Reticketed upgrading the payday snapshot as #3444.

**Comment by whit537** [comment]
MassPay done and posted back for two users.

**Comment by whit537** [comment]
Not gonna worry about shuffling this week, since amounts were so small.

**Comment by whit537** [comment]
Logs uploaded in https://github.com/gratipay/logs/commit/5330cb28bf936446c144174b534214a382519016 and https://github.com/gratipay/logs/commit/ae4188f357cf664653d5ca04e177073c81045629.

**Comment by whit537** [comment]
Okay! *Now* we're done. :-)

:birthday:

**Comment by whit537** [comment]
:sigh:

Not seeing MassPay in `exchanges` ...

**Comment by whit537** [comment]
Fixed in https://github.com/gratipay/gratipay.com/commit/b48db583db87bee852e56122b9ee9611ed0e9a87 and postback completed and confirmed. Okay!

!m @rohitpaulk 
!m *

**Comment by rohitpaulk** [comment]
Noticed this when working on https://github.com/gratipay/gratipay.com/issues/3705

Comparing https://github.com/gratipay/gratipay.com/issues/3415#issuecomment-102488894 and https://github.com/gratipay/gratipay.com/issues/3415#issuecomment-102508212... 

`Payday started at 2015-05-15 18:42:11.415296+00:00.`
and
`Payday started at 2015-05-14 21:34:37.995735+00:00`

We somehow messed up the year here (2015 changed to 2014)... I'm going to manually correct this in the DB.

**Comment by rohitpaulk** [comment]
```sql
=> SELECT ts_start, ts_end FROM paydays WHERE id=199;
           ts_start            |            ts_end
-------------------------------+-------------------------------
 2015-05-14 21:34:37.995735+00 | 2015-05-15 20:13:08.453174+00

=> UPDATE paydays SET ts_start='2015-05-15 18:42:11.415296+00:00' WHERE id=199;
UPDATE 1

=> SELECT ts_start, ts_end FROM paydays WHERE id=199;
           ts_start            |            ts_end
-------------------------------+-------------------------------
 2015-05-15 18:42:11.415296+00 | 2015-05-15 20:13:08.453174+00
(1 row)
```

**Comment by whit537** [comment]
> We somehow messed up the year here (2015 changed to 2014)... I'm going to manually correct this in the DB.

@rohitpaulk I'm seeing this in email notification but not here. What's the story?

**Comment by rohitpaulk** [comment]
>  I'm seeing this in email notification but not here. What's the story?

Ah that was me misreading data - I tried it out on a backup and then realized the mistake. Apparently, my attempts to cover the history didn't work as planned :stuck_out_tongue:

**Comment by whit537** [comment]
:-)

**Comment by whit537** [comment]
We seem to be making progress on https://github.com/gratipay/gratipay.com/issues/3415#issuecomment-102178928 over in https://github.com/gratipay/inside.gratipay.com/issues/477.
